### ISSUE TITLE
IDs for Receptor objects

### ISSUE THREAD

**Issue by bussec** [airr-community/airr-standards#540]
_Originally posted by @schristley in https://github.com/airr-community/airr-standards/issues/524#issuecomment-868818350_

IEDB has [released](https://help.iedb.org/hc/en-us/articles/4402872882189-Immune-Epitope-Database-Query-API-IQ-API-) a beta query API for their database.

Of interest is they have a [curie_map endpoint](https://query-api.iedb.org/curie_map) which returns a structure similar to our `CURIEMap`. I've noticed however that the IRI goes to the human-readable html page instead of the API which returns machine-readable JSON. For example, `IEDB_EPITOPE: 7355` resolves to:

```
https://www.iedb.org/epitope/7355
```

versus

```
https://query-api.iedb.org/epitope_search?structure_id=eq.7355
```

The nice thing is that now we should be to add a single field in Rearrangement if we want to link to an epitope #44 

Though that opens the question on whether we should put entries into our CURIEMap for resolving `IEDB_EPITOPE` or if we should use IEDB's?

We also can consider how we might link with receptors. The IEDB API has `tcr_search` and `bcr_search` endpoints:

```
https://query-api.iedb.org/tcr_search?limit=5
https://query-api.iedb.org/bcr_search?limit=5
```

The receptor ids have `IEDB_RECEPTOR` as their CURIE though oddly it's missing from the above curie_map. However, we might want to consider how we can link our `Receptor` to IEDB's.

**Comment by bussec** [comment]
> The receptor ids have IEDB_RECEPTOR as their CURIE though oddly it's missing from the above curie_map. However, we might want to consider how we can link our Receptor to IEDB's.

My assumption would be that it  is not in the `curie_map` because IEDB does not offer a separate endpoint to query their *Receptor Group* objects. Looking at the IEDB SQL database dump, getting this information would require several JOINs as database tables are by and large normalized:
```
receptor_group <-- distinct_receptor_receptor_grp --> distinct_receptor <-- curated_receptor --> chain
                                                                        \-> distinct_chain
```
Notes:
*  Arrows point from the table with a foreign key to the table containing the primary key
* The tables `chain` and `distinct_chain` share the layout (incl.: `v_dom_seq`, `[vdj]_gene`, `cdr[1-3]_(seq|start|end)`), with mainly `[vdj]_gene` and `cdr3_seq` being populated (i.e., not NULL).
* `distinct_chain` seems only to use IMGT designations, while `chain` can have two entries for an observation, with different gene designations.
* `curated_receptor` links to separate `_cur` und `_calc` `chains`, those are the once with the different gene calls

So it should not be that difficult to get the receptor information internally and `receptor_group` would be suited to represent multiple observations of functionally identical receptors.

The thing that puzzles me, however, is that I could not find a table that would contains the complete sequence for each chain (AA or nucleic acid). The information is occasionally provided in the `v_dom_seq` field. So it appears to me as if IEDB reduces this information to the `v_call`. While this is legitimate, especially for TCRs, but it would create two issues for us:

1. Our idea to hash the AA sequence of both receptor chains would require a reverse lookup of the sequence, which would fail in the case of SHMs or unresolved SNPs.
2. The system is very sensitive to renaming of genes/alleles, as the new designation cannot be recalculated.

We therefore should clarify with IEDB whether a hash-based ID would work or not...

**Comment by schristley** [comment]
> The thing that puzzles me, however, is that I could not find a table that would contains the complete sequence for each chain (AA or nucleic acid). The information is occasionally provided in the `v_dom_seq` field. So it appears to me as if IEDB reduces this information to the `v_call`. While this is legitimate, especially for TCRs, but it would create two issues for us:

I think the complete sequence information isn't always available from the curated publication. For example, I picked a random receptor and looked at the [publication](https://pubmed.ncbi.nlm.nih.gov/16474394/). It doesn't provide any sequence but instead points to PDB for [2ESV](https://www.rcsb.org/sequence/2ESV) and [1MI5](https://www.rcsb.org/sequence/1MI5), which has the sequence available in a FASTA download format, but it only seems to be AA. Not sure if nucleotide sequence is available at all.

**Comment by schristley** [comment]
> My assumption would be that it is not in the `curie_map` because IEDB does not offer a separate endpoint to query their _Receptor Group_ objects.

Not a receptor group endpoint, but they do have the `tcr_search` and `bcr_search` endpoints. The data returned from that endpoint should be similar to what is exported from the website, and has fields for complete AA sequences, though they could be null.

**Comment by bcorrie** [comment]
We are using the IEDB query API to provide the following functionality on the iReceptor Gateway when you search a junction:

- It searches the ADC repositories and gives you the normal search result.
- It searches IEDB and if it finds any records it displays the following (ignore the error, one of our test DBs is done)

<img width="739" alt="image" src="https://user-images.githubusercontent.com/1105216/155467041-67076a49-5729-4e92-bd3b-a4c2ad57c616.png">

It works quite well. In fact, if you search for a Junction and it finds it in IEDB but not the ADC it still gives you the IEDB info. We are almost ready to put this into production.

**Comment by bcorrie** [comment]
> Not a receptor group endpoint, but they do have the `tcr_search` and `bcr_search` endpoints.

There is no singular receptor endpoint true, but according to the their docs (https://help.iedb.org/hc/en-us/articles/4402872882189-Immune-Epitope-Database-Query-API-IQ-API-) the tcr_search and bcr_search endpoints are their receptor endpoints. These entry point returns an array of receptor objects.

You can search for a IEDB_RECEPTOR CURIE using the TCR and BCR searches as follows:

https://query-api.iedb.org/tcr_search?receptor_group_iri=eq.IEDB_RECEPTOR:8670
or
https://query-api.iedb.org/tcr_search?receptor_group_id=eq.8670

You can get to the equivalent human friendly page here:

https://www.iedb.org/receptor/8670

So this seems like it might make sense to add these to CURIEMap and InformationProvider, assuming we do want to link to them from the Receptor object.

**Comment by bussec** [comment]
CURIE mapping for `IEDB_RECEPTOR` to `https://www.iedb.org/receptor/` has now been included in PR #404. Referencing to receptor groups in IEDB is possible via the `receptor_ref` array in the `Receptor` object. Furthermore, while the receptor object does currently not directly support IEDB epitopes, it uses the same ontologies for antigens, thereby ensuring semantic interoperability.

This issue is set to auto-close upon merging of #404.
